<a href="https://colab.research.google.com/github/sarumathicsc/flyrank-ml-journey/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Sagnik556/flyrank-ml-internship-starter/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

# 1. My Lane and Why

## Provisional Lane: Growth / Recovery / Momentum Prediction

I am exploring the Growth / Recovery / Momentum Prediction lane. My goal is to investigate whether historical search-performance signals can help identify content pages that may experience a significant future decline in search performance.

I chose this lane because my earlier experiments showed that search-performance features and query-level signals can contain useful predictive information. In Week 3, I also experimented with position volatility and client-based evaluation.

This is a provisional lane, and I may refine the research question as I learn more about the warehouse data.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

# 2. The Question: Decision, Action, and Cost

## Research Question

Can historical search performance, position stability, and query-level signals help identify content pages that are likely to experience a significant decline in search impressions during a future time window?

## Unit of Analysis

The primary unit of analysis will be a content page at a specific decision point.

## Decision

The decision is which content pages should be reviewed first because they may be at risk of future search-performance decline.

## Output

The final output will be a ranked list of content pages with a predicted decline-risk score and interpretable reason codes explaining the signals behind each recommendation.

## Action

A content or SEO reviewer could investigate the highest-ranked pages and decide whether they require actions such as content updates, search-performance analysis, or further review.

## Cost of a Wrong Recommendation

A false positive could cause a reviewer to spend limited time investigating a page that does not actually decline. A false negative could cause an important declining page to be overlooked.

For this reason, the system should be treated as decision support rather than a guarantee that a page will decline or recover.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [6]:
import pandas as pd

url = "https://raw.githubusercontent.com/sarumathicsc/flyrank-ml-journey/main/data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(url)

df.head()

,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [7]:
print("Total content pages:", len(df))

declining = (df["trend_direction"] == "down").sum()
declining_pct = declining / len(df) * 100

print("Pages with declining performance:", declining)
print(f"Percentage of declining pages: {declining_pct:.2f}%")

print("\nAverage impressions_90d by trend direction:")
print(df.groupby("trend_direction")["impressions_90d"].mean().round(2))

print("\nNumber of unique clients:", df["client_id"].nunique())

Total content pages: 30000
Pages with declining performance: 16262
Percentage of declining pages: 54.21%

Average impressions_90d by trend direction:
trend_direction
down      4919.10
flat        20.74
new        164.13
stable    9213.62
up        4716.08
Name: impressions_90d, dtype: float64

Number of unique clients: 32


### Observations from the Starter Dataset

The starter dataset contains **30,000 content pages across 32 clients**. I observed that **16,262 pages (54.21%)** are classified with a downward trend, suggesting that search-performance decline is a common enough pattern to investigate.

The average 90-day impressions also differ across trend groups. Pages with a **stable** trend have an average of **9,213.62 impressions**, while pages with a **downward** trend average **4,919.10 impressions**. This observed difference suggests that historical search-performance patterns may contain useful directional signals for identifying pages that could require attention.

These numbers support my provisional **Growth / Recovery / Momentum Prediction** lane. However, these observations do not prove that any individual feature causes a decline. The goal of the project is to provide **decision-support** by identifying potentially at-risk content pages for further review.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

#Careful Words: What I Can and Cannot Claim

This project can investigate whether patterns in historical data are associated with future changes in search performance.

I can evaluate whether a model predicts future decline better than a baseline on held-out data and previously unseen clients.

However, I cannot claim that a specific feature causes a page to decline. I also cannot claim that updating a recommended page will automatically cause its search performance to recover.

The final system will therefore be a decision-support tool that helps prioritize pages for human review.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.